ext/mini_racer_extension/mini_racer_extension.c

```diff
@@ -0,0 +1,1449 @@
+#include <stdatomic.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <pthread.h>
+
+#include "ruby.h"
+#include "ruby/encoding.h"
+#include "ruby/thread.h"
+#include "serde.c"
+#include "mini_racer_v8.h"
+
+#define countof(x)  (sizeof(x) / sizeof(*(x)))
+#define endof(x)    ((x) + countof(x))
+
+// work around missing pthread_barrier_t on macOS
+typedef struct Barrier
+{
+    pthread_mutex_t mtx;
+    pthread_cond_t cv;
+    int count, in, out;
+} Barrier;
+
+static inline int barrier_init(Barrier *b, int count)
+{
+    int r;
+
+    if ((r = pthread_mutex_init(&b->mtx, NULL)))
+        return r;
+    if ((r = pthread_cond_init(&b->cv, NULL))) {
+        pthread_mutex_destroy(&b->mtx);
+        return r;
+    }
+    b->count = count;
+    b->out = 0;
+    b->in = 0;
+    return 0;
+}
+
+static inline void barrier_destroy(Barrier *b)
+{
+    pthread_mutex_destroy(&b->mtx);
+    pthread_cond_destroy(&b->cv);
+}
+
+static inline int barrier_wait(Barrier *b)
+{
+    int last;
+
+    pthread_mutex_lock(&b->mtx);
```

Comment:
should we be checking return values?

Reply:
They can't really error, they're not PTHREAD_MUTEX_ERRORCHECK locks; logic errors result in deadlock. I could call abort() if you want.